Add aws-lambda cd tutorial (#418)
timescale/docs: timescaledb/tutorials/aws-lambda/continuous-deployment.md

You can use GitHub actions to set up automatic deployment for AWS Lambda from a Github repository.
You need to push a commit to the main or master branch of your repository, then let GitHub actions create the deployment 
package, and deploy your code to AWS Lambda.

## Prerequisites
* Git ([installation options here](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git))
* GitHub CLI tool ([installation options here](https://github.com/cli/cli#installation))
* AWS account

## Procedure: Creating a new Lambda function in AWS console
1.  Create a new Lambda function called `lambda-cd` by navigating to the AWS Lambda console and creating a new function:
    ![create new function](https://assets.timescale.com/docs/images/tutorials/aws-lambda-tutorial/create_new_function.png)
1.  Add `lambda-cd` as your function name, choose your preferred runtime environment, and click `Create function`:
    ![create lambda from scratch](https://assets.timescale.com/docs/images/tutorials/aws-lambda-tutorial/from_scratch.png)
    Take a note of the function name, you need it later to configure the deployment file.

## Procedure: Creating a new GitHub repository

> 
> Now let's create a new GitHub repository which contains the function code.

Reviewer: ```suggestion
Now you can create a new GitHub repository which contains the function code.
```

1.  Create a new GitHub [repository](https://github.com/new).
    ![lambda github repo](image url)

> 1.  On your computer, create a new project folder called `lambda-cd`:

Reviewer: ```suggestion
1.  On your local system, create a new project folder called `lambda-cd`:
```

    import psycopg2
    import psycopg2.extras
    import os
    
    def lambda_handler(event, context):
        db_name = os.environ['DB_NAME']
        db_user = os.environ['DB_USER']
        db_host = os.environ['DB_HOST']
        db_port = os.environ['DB_PORT']
        db_pass = os.environ['DB_PASS']
        
        conn = psycopg2.connect(user=db_user, database=db_name, host=db_host,
                                password=db_pass, port=db_port)
        
        sql = "SELECT * FROM stocks_intraday ORDER BY time DESC LIMIT 10"
        cursor = conn.cursor(cursor_factory=psycopg2.extras.DictCursor)
        cursor.execute(sql)
        result = cursor.fetchall()
        
        return {
            'statusCode': 200,
            'body': json.dumps(list_of_dicts, default=str),
            'headers': {
                "Content-Type": "application/json"
            }
        }
    ```
1.  Initialize a git repository and push the project to GitHub:
    ```bash
    git init
    git add function.py
    git commit -m "Initial commit: add Lambda function"
    git branch -M master
    git remote add origin <YOUR_GITHUB_PROJECT_URL.git>
    git push -u origin master
    ```

> At this point, you have a GitHub repo with just the Lambda function in it. Let's see how to connect this repo to the AWS Lambda function.

Reviewer: ```suggestion
At this point, you have a GitHub repository with just the Lambda function in it. Now you can connect this repository to the AWS Lambda function.
```

## Connect GitHub and AWS Lambda

> 

Reviewer: This needs a sentence.

Author: https://github.com/timescale/docs/pull/418/commits/dbee9ec713e0d08e4a7080750b46ce46e587925e

### Procedure: Adding your AWS credentials to the repository
You need to add your AWS credentials to the repository so it will have permission to connect to Lambda. You can do this
by adding [GitHub secrets](https://docs.github.com/en/actions/reference/encrypted-secrets) using the GitHub CLI.

> 1.  Authenticate with a GitHub:

Reviewer: ```suggestion
1.  Authenticate with GitHub:
```

    ```bash
    gh auth login
    ```

>     This will prompt you to choose what account you want to log into using either your password or GitHub

Reviewer: ```suggestion
    This prompts you to choose which account you want to log into using either your password or GitHub
```

    authentication token.
1.  Add AWS credentials as GitHub secrets. 
    By using GitHub secrets, your credentials are encrypted and cannot be seen 
    publicly. Use the `gh secret set` command to upload your AWS credentials one by one 
    (you'll be prompted to paste the values for each one):
    
    AWS_ACCESS_KEY_ID:
    ```bash
    gh secret set AWS_ACCESS_KEY_ID
    ```
    
    AWS_SECRET_ACCESS_KEY:
    ```bash
    gh secret set AWS_SECRET_ACCESS_KEY
    ```
    
    AWS_REGION:
    ```bash
    gh secret set AWS_REGION
    ```
    

>     To make sure your credentials have been uploaded correctly, you can list the available GitHub secrets:

Reviewer: ```suggestion
1.   To make sure your credentials have been uploaded correctly, you can list the available GitHub secrets:
```

    ```bash
    gh secret list
    AWS_ACCESS_KEY_ID      Updated 2021-09-13
    AWS_SECRET_ACCESS_KEY  Updated 2021-09-13
    AWS_REGION             Updated 2021-09-13
    ```

Now you know that you have your AWS credentials available for the repository to use.

### Procedure: Setting up GitHub actions

> You will set up a new automation based on the ["AWS Lambda Deploy" GitHub action](https://github.com/marketplace/actions/aws-lambda-deploy)

Reviewer: ```suggestion
You can now set up some automation based on the ["AWS Lambda Deploy" GitHub action](https://github.com/marketplace/actions/aws-lambda-deploy)
```

to auto-deploy to AWS Lambda.

> 1.  Create a new YAML file (this will contain the deployment configuration):

Reviewer: ```suggestion
1.  Create a new YAML file that contains the deployment configuration:
```

    ```bash
    touch .github/workflows/main.yml
    ```
    

>     And add this to the file:

Reviewer: ```suggestion
1.  Add this content to the file:
```

    ```yml
    name: deploy to lambda
    on:
      # Trigger the workflow on push or pull request,
      # but only for the main branch
      push:
        branches:
          - master
    jobs:
    
      deploy_source:
        name: deploy lambda from source
        runs-on: ubuntu-latest
        steps:
          - name: checkout source code
            uses: actions/checkout@v1
          - name: default deploy
            uses: appleboy/lambda-action@master
            with:
              aws_access_key_id: ${{ secrets.AWS_ACCESS_KEY_ID }}
              aws_secret_access_key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
              aws_region: ${{ secrets.AWS_REGION }}
              function_name: lambda-cd
              source: function.py
    ```

>     This configuration will make sure to deploy the code to Lambda when there's a new push to the master branch.

Reviewer: ```suggestion
    This configuration ensures that the code is deployed to Lambda when there's a new push to the `master` branch.
```

Shouldn't we be using `main` now?

Author: Does it make a difference? I've been using master throughout each AWS tutorial, so I kept the same here. Maybe we should do a search and replace on each Lambda tutorial to master->main?

Reviewer: GH have updated the default to `main`, in a general move away from language that could be considered insensitive. See also https://inclusivenaming.org/. I think we probably should do a search through, yes.

Author: to be fixed here https://github.com/timescale/docs/issues/456

    As you can also see in the YAML file, the AWS credentials are accessed using the `${{ secrets.AWS_ACCESS_KEY_ID }}` syntax.

> 1.  Make sure the use the same function name in this configuration file as what's the name of the Lambda function in

Reviewer: This doesn't seem like a new step. It needs to be attached to the previous step, or this whole end section moved out of the procedure.

    the AWS console ("lambda-cd" in this example).

### Procedure: Testing the pipeline

> 
> You can test if the hook works by pushing the changes.

Reviewer: ```suggestion
You can test if the hook works by pushing the changes to GitHub.
```

> 1.  Push the changes to the repo.

Reviewer: ```suggestion
1.  Push the changes to the repository:
```

    ```bash
    git add .github/workflows/main.yml
    git commit -m "Add deploy configuration file"
    git add function.py
    git commit -m "Update function code"
    git push
    ```
1.  Navigate to the GitHub actions page of your repository, to see the build run and succeed:
    ![github action build](https://assets.timescale.com/docs/images/tutorials/aws-lambda-tutorial/github_action_lambda.png)

> With this step completed, you have a continuous deployment pipeline set up between GitHub and AWS Lambda.

Reviewer: ```suggestion
You now have a continuous deployment pipeline set up between GitHub and AWS Lambda.
```

> # Lambda continuous deployment with Github Actions
> 

Reviewer: ```suggestion
# Lambda continuous deployment with GitHub actions
```

> This tutorial builds a continuous deployment (CD) pipeline between Github and AWS Lambda using Github Actions.

Reviewer 2: ```suggestion
This tutorial builds a continuous deployment (CD) pipeline between GitHub and AWS Lambda using GitHub Actions.
```
Really, anywhere we reference it, this is the proper capitalization. You can do the rest. 😃

Reviewer: Also, `actions` should be lowercased.

Author: https://github.com/timescale/docs/pull/418/commits/58f1063e6f0828653ca36fed25de09b63b54bbae

> When using AWS Lambda, packaging and deploying your function and its dependencies can be a tedious job. What if you
> also want to use a source code management platform like Github to develop your code before pushing it to AWS Lambda?

Reviewer: ```suggestion
Packaging and deploying your function and its dependencies with AWS Lambda can sometimes be a tedious job. Especially if you also want to use a source code management platform like GitHub to develop your code before pushing it to AWS Lambda.
```

> 
> Luckily, you can use Github actions to set up automatic deployment for AWS Lambda from a Github repository.
> You just need to push a commit to your main/master branch of your repository then let Github Actions do the heavy
> lifting, creating the deployment package and deploying your code to AWS Lambda.

Reviewer: ```suggestion
You can use GitHub actions to set up automatic deployment for AWS Lambda from a Github repository.
You need to push a commit to the main or master branch of your repository, then let GitHub actions create the deployment package, and deploy your code to AWS Lambda.
```

Author: https://github.com/timescale/docs/pull/418/commits/fd43ee2a9b8ea3ef467bc80be194210017b2f61f

## Prerequisites
* Git ([installation options here](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git))

> * Github CLI tool ([installation options here](https://github.com/cli/cli#installation))

Reviewer: ```suggestion
* GitHub CLI tool ([installation options here](https://github.com/cli/cli#installation))
```

* AWS account

> ## Create a new Lambda function in AWS console
> 
> Create a new Lambda function called "lambda-cd"
> Navigate to the AWS Lambda console and create a new function:
> ![create new function](https://assets.timescale.com/docs/images/tutorials/aws-lambda-tutorial/create_new_function.png)
> 
> 
> Add "lambda-cd" as your function name and choose your preferred runtime environment:
> 
> ![create lambda from scratch](https://assets.timescale.com/docs/images/tutorials/aws-lambda-tutorial/from_scratch.png)
> 
> 
> After clicking "Create function", you have created a Lambda function called "lambda-cd". Remember the name of the function because later you'll need to reference this
> to configure the deployment file.

Reviewer: ```suggestion
## Procedure: Creating a new Lambda function in AWS console
1.  Create a new Lambda function called `lambda-cd` by navigating to the AWS Lambda console and creating a new function:
    ![create new function](https://assets.timescale.com/docs/images/tutorials/aws-lambda-tutorial/create_new_function.png)
1.  Add `lambda-cd` as your function name, choose your preferred runtime environment, and click `Create function`:
    ![create lambda from scratch](https://assets.timescale.com/docs/images/tutorials/aws-lambda-tutorial/from_scratch.png)
    Take a note of the function name, you need it later to configure the deployment file.
```

Reviewer: You need to do this for all the following procedures as well 👇
